File: src/ipc/events.ts
/**
 * Event-subscription helpers for the PTY session event stream.
 *
 * Each helper subscribes to a Tauri event and returns a promise resolving to an
 * unlisten function; call it to tear down the subscription.
 */
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

import type { ExitEvent, OutputEvent, UsageUpdate } from "../types";

/** Subscribe to `session://output`; returns an unlisten function. */
export function onSessionOutput(
  cb: (payload: OutputEvent) => void,
): Promise<UnlistenFn> {
  return listen<OutputEvent>("session://output", (event) => cb(event.payload));
}

/** Subscribe to `session://exit`; returns an unlisten function. */
export function onSessionExit(
  cb: (payload: ExitEvent) => void,
): Promise<UnlistenFn> {
  return listen<ExitEvent>("session://exit", (event) => cb(event.payload));
}

/**
 * Subscribe to `commander://navigate`; returns an unlisten function. The
 * payload is the bare directory path string Commander resolved, used to select
 * that directory in the sidebar.
 */
export function onCommanderNavigate(
  cb: (path: string) => void,
): Promise<UnlistenFn> {
  return listen<string>("commander://navigate", (event) => cb(event.payload));
}

/**
 * Subscribe to `commander://directories-changed`; returns an unlisten function.
 * Fired when a Commander MCP tool (`add_project` / `remove_project`) mutates the
 * managed directory list — the sidebar reloads it so the change shows up live
 * instead of only in the persisted store. The payload is empty.
 */
export function onCommanderDirectoriesChanged(
  cb: () => void,
): Promise<UnlistenFn> {
  return listen("commander://directories-changed", () => cb());
}

/**
 * Subscribe to `usage://updated`; returns an unlisten function. Fired by the
 * backend meter whenever the rolling-window snapshot changes. The payload is a
 * `UsageUpdate` carrying either fresh data or an error string.
 */
export function onUsageUpdated(
  cb: (u: UsageUpdate) => void,
): Promise<UnlistenFn> {
  return listen<UsageUpdate>("usage://updated", (event) => cb(event.payload));
}

/**
 * Subscribe to `session://paused`; returns an unlisten function. Fired when a
 * session is OS-suspended/resumed (by Commander or the tile resume control).
 */
export function onSessionPaused(
  cb: (payload: { sessionId: string; paused: boolean }) => void,
): Promise<UnlistenFn> {
  return listen<{ sessionId: string; paused: boolean }>(
    "session://paused",
    (event) => cb(event.payload),
  );
}

File: src-tauri/src/events.rs
//! Event-name constants and payload structs emitted to the frontend.
//!
//! Payload field names are serialized in camelCase to match the TypeScript
//! mirror (`ts-ipc-contract`).

use serde::{Deserialize, Serialize};

use crate::usage::UsageSnapshot;

/// Emitted with a chunk of session stdout/stderr output.
pub const SESSION_OUTPUT: &str = "session://output";

/// Emitted when a session's PTY process exits.
pub const SESSION_EXIT: &str = "session://exit";

/// Emitted when the Commander's `resolve_project` MCP tool matches a managed
/// directory: the UI should select that directory. Payload is the directory
/// path as a bare JSON string. The frontend (`commander-app-wiring`) listens on
/// this exact name.
pub const COMMANDER_NAVIGATE: &str = "commander://navigate";

/// Emitted when a Commander MCP tool mutates the managed directory list
/// (`add_project` / `remove_project`): the sidebar should reload it via
/// `list_directories`. Without this, a Commander-driven add/remove only lands
/// in the persisted store and the frontend's once-loaded list stays stale until
/// the next mount. Payload is empty (unit). The frontend listens on this name.
pub const COMMANDER_DIRECTORIES_CHANGED: &str = "commander://directories-changed";

/// Emitted when a session is OS-suspended or resumed by Commander (or the UI
/// resume control). Carries the session id and its new paused state.
pub const SESSION_PAUSED: &str = "session://paused";

/// Emitted by the background usage poller on each refresh of the Claude
/// subscription usage limits: immediately on startup, then on a fixed cadence.
/// Carries a [`UsageUpdate`] — a fresh [`UsageSnapshot`] on success or an error
/// string on failure. The frontend listens on this exact name.
pub const USAGE_UPDATED: &str = "usage://updated";

/// Payload for [`SESSION_OUTPUT`].
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputEvent {
    pub session_id: String,
    pub data: String,
}

/// Payload for [`SESSION_EXIT`].
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ExitEvent {
    pub session_id: String,
    pub code: Option<i32>,
}

/// Payload for [`SESSION_PAUSED`].
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PausedEvent {
    pub session_id: String,
    pub paused: bool,
}

/// Payload for [`USAGE_UPDATED`].
///
/// Exactly one of `snapshot` / `error` is set: a fresh [`UsageSnapshot`] with
/// `error` `None` on a successful poll, or `snapshot` `None` with the error
/// string on a failed one.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UsageUpdate {
    pub snapshot: Option<UsageSnapshot>,
    pub error: Option<String>,
}
